### PR TITLE
KCORE-440: add voter information towards the quorum-state file

### PR DESCRIPTION
**Pull request by abbccdda**
Amend a gap for quorum state information to add the known voters into the state file.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java [hunk 1]
```diff
@@ -62,8 +62,8 @@ public void initialize(OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
             log.warn("Clear local quorum state store {}", store.toString(), e);
             store.clear();
 
-            election = ElectionState.withUnknownLeader(0);
-            state = new FollowerState(election.epoch);
+            election = ElectionState.withUnknownLeader(0, voters);
```
Comment (hachikuji) [inline raft/src/main/java/org/apache/kafka/raft/QuorumState.java:65]: In the case that the state is successfully loaded from the file, should we validate that it matches the configured set of voters?

File: raft/src/test/java/org/apache/kafka/raft/RaftEventSimulationTest.java [hunk 2]
```diff
@@ -129,13 +129,16 @@ public void testReplicationNoLeaderChangeQuorumSizeFive() throws IOException {
     }
 
     private void testReplicationNoLeaderChange(QuorumConfig config) throws IOException {
+        Set<Integer> voters = getVoters(config.numVoters);
```
Comment (hachikuji) [inline raft/src/test/java/org/apache/kafka/raft/RaftEventSimulationTest.java:132]: Do we need this? `Cluster` already has a method to get the set of voters.

Comment (abbccdda) [reply raft/src/test/java/org/apache/kafka/raft/RaftEventSimulationTest.java:132]: Yea, good call!

File: raft/src/main/java/org/apache/kafka/raft/QuorumState.java (hunk 1 above)
Comment (abbccdda) [reply raft/src/main/java/org/apache/kafka/raft/QuorumState.java:65]: Good question, I feel we should fail the initialization for the first version as the state file takes precedence and user should always ensure to update the configuration for all or nothing.

File: raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java [hunk 3]
```diff
@@ -155,8 +159,8 @@ public void testCannotBecomeLeaderIfAlreadyLeader() throws IOException {
     public void testCannotBecomeFollowerOfSelf() throws IOException {
         Set<Integer> voters = Utils.mkSet(localId);
         assertEquals(0, store.readElectionState().epoch);
-        QuorumState state = new QuorumState(localId, voters, store, new LogContext());
-        state.initialize(new OffsetAndEpoch(0L, logEndEpoch));
+        QuorumState state = initializeState(voters, store);
+//        state.initialize(new OffsetAndEpoch(0L, logEndEpoch));
```
Comment (hachikuji) [inline raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:163]: I guess we don't need this?

File: raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java [hunk 4]
```diff
@@ -204,37 +208,53 @@ public void testCannotBecomeLeaderWithoutGrantedVote() throws IOException {
...
         state.candidateStateOrThrow().recordGrantedVote(otherNodeId);
         state.becomeLeader(0L);
         assertTrue(state.becomeFetchingFollower(5, otherNodeId));
         assertEquals(5, state.epoch());
         assertEquals(OptionalInt.of(otherNodeId), state.leaderId());
-        assertEquals(ElectionState.withElectedLeader(5, otherNodeId), store.readElectionState());
+        assertEquals(ElectionState.withElectedLeader(5, otherNodeId, voters), store.readElectionState());
+    }
+
+    private QuorumState initializeState(Set<Integer> voters,
```
Comment (hachikuji) [inline raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:223]: To be clearer, could could we call this `initializeEmptyState` or something like that?

Comment (abbccdda) [reply raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:223]: Make sense!

**Pull request merged**